cryptarchia/cryptarchia.py

```diff
@@ -118,37 +122,103 @@ def on_block(self, block: BlockHeader):
         return self.fork_choice()
 
 
 @dataclass
 class Coin:
+    pk: int
     value: int
 
+
+@dataclass
+class LedgerState:
+    """
+    A snapshot of the ledger state up to some height
+    """
+
+    height: int = None
+    nonce: bytes = None
+    total_stake: int = None
+
+
+@dataclass
+class EpochState:
+    # for details of snapshot schedule please see:
+    # https://github.com/IntersectMBO/ouroboros-consensus/blob/fe245ac1d8dbfb563ede2fdb6585055e12ce9738/docs/website/contents/for-developers/Glossary.md#epoch-structure
+
+    # The stake distribution snapshot is taken at the beginning of the previous epoch
+    stake_distribution_snapshot: LedgerState
+
+    # The nonce snapshot is taken 7k/f slots into the previous epoch
+    nonce_snapshot: LedgerState
+
+    def stake_distribution(self) -> int:
+        """Returns the total stake that will be used to reletivize leadership proofs during this epoch"""
+        return self.stake_distribution_snapshot.total_stake
+
+    def nonce(self) -> bytes:
+        return self.nonce_snapshot.nonce
+
+
+@dataclass
+class LeaderConfig:
+    active_slot_coeff: float = 0.05  # 'f', the rate of occupied slots
+
+
+def phi(f: float, alpha: float) -> float:
+    """
+    params:
+      f: 'active slot coefficient' - the rate of occupied slots
+      alpha: relative stake held by the validator
+
+    returns: the probability that this validator should win the slot lottery
+    """
+    return 1 - (1 - f) ** alpha
+
+
+class MOCK_LEADER_VRF:
+    """NOT SECURE: A mock VRF function where the sk and pk are assummed to be the same"""
+
+    ORDER = 2**256
+
+    @classmethod
+    def vrf(cls, sk: int, nonce: bytes, slot: int) -> int:
+        h = sha256()
+        h.update(int.to_bytes(sk, length=32))
+        h.update(nonce)
+        h.update(int.to_bytes(slot, length=16))  # 64bit slots
+        return int.from_bytes(h.digest())
+
+    @classmethod
+    def verify(cls, r, pk, nonce, slot):
+        raise NotImplemented()
+
+
+@dataclass
 class Leader:
-    def init(self, genesis: BlockHeader, config: TimeConfig, coins: List[Coin]):
-        self.config = config
-        self.tip = genesis
-        self.coins = coins
+    config: LeaderConfig
+    coin: Coin
+
+    def is_slot_leader(self, epoch: EpochState, slot: Slot) -> bool:
+        f = self.config.active_slot_coeff
+        total_stake = epoch.stake_distribution()
+        relative_stake = self.coin.value / total_stake
+
+        r = MOCK_LEADER_VRF.vrf(self.coin.pk, epoch.nonce(), slot)
+
+        return r < MOCK_LEADER_VRF.ORDER * phi(f, relative_stake)
 
-    def is_leader_at(self, slot: Slot) -> bool:
-        for coin in self.coins:
-            if lottery(slot, coin):
-                return True
-        return False
-    
     def propose_block(self, slot: Slot, parent: BlockHeader) -> BlockHeader:
         assert self.is_leader_at(slot)
         return BlockHeader(parent=parent.id(), slot=slot, _id=Id(b"TODO"))
```

Comment:
`_id` has been removed?

Reply:
Right, let me remove that line, something about the underscore in `_id` was not playing well with python dataclasses.

Besides, I think block ID should be a derived field of the block header.

Reply:
> Besides, I think block ID should be a derived field of the block header.

yes, it's just that we don't have enough data for now